lib/Conversion/TorchToLinalg/AdaptivePooling.cpp

```diff
@@ -0,0 +1,237 @@
+    RankedTensorType InputType = input.getType().cast<RankedTensorType>();
+    RankedTensorType OutputType =
+        typeconverter->convertType(op.getResult().getType())
+            .cast<RankedTensorType>();
+
+    // get rank of input (same as rank of output)
+    int64_t rank = input.getType().cast<RankedTensorType>().getRank();
+    // input operand should be NCH (i.e. rank 3)
+    if (rank != 3) {
+      return rewriter.notifyMatchFailure(op, "only supports input type NCH");
+    }
+
+    // get elementType of input tensor
+    Type elementType = InputType.getElementType();
+
+    // make an iteration space of size Kmax = 1 + ceildiv (Hin - 1) , Hout
+    Type dummyType = rewriter.getI1Type();
+    Value Kiter;
+    if (InputType.isDynamicDim(rank - 1) || OutputType.isDynamicDim(rank - 1)) {
+      Value constantOne =
+          rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(1));
+      Value HinPlusOne = rewriter.create<arith::SubIOp>(loc, Hin, constantOne);
+      Value KmaxMinusOne =
+          rewriter.create<arith::CeilDivSIOp>(loc, HinPlusOne, HoutIndex);
+      Value Kmax =
+          rewriter.create<arith::AddIOp>(loc, constantOne, KmaxMinusOne);
+      Kiter = rewriter.create<tensor::EmptyOp>(
+          loc, RankedTensorType::get({ShapedType::kDynamic}, dummyType), Kmax);
+    } else {
+      int64_t HinInt = InputType.getShape()[rank - 1];
+      int64_t HoutInt = OutputType.getShape()[rank - 1];
+      int64_t correction = ((HinInt - 1) % HoutInt == 0) ? 0 : 1;
+      int64_t Kmax = 1 + (HinInt - 1) / HoutInt + correction;
+      Kiter = rewriter.create<tensor::EmptyOp>(
+          loc, RankedTensorType::get({Kmax}, dummyType), ValueRange{});
+    }
+
+    // need to buffer input, else there will possibly be an out of bounds access
+    // later BuffVal = 0 for avg pooling and -inf for max pooling
+    Value BuffVal = rewriter.create<arith::ConstantOp>(
+        loc, elementType, rewriter.getFloatAttr(elementType, 0));
+    SmallVector<int64_t> lowPadding = {0, 0, 0};
+    SmallVector<int64_t> highPadding = {0, 0, 1};
+    Value BuffInput = torch_to_linalg::getPaddedTensor(
+        op, rewriter, input, lowPadding, highPadding, BuffVal);
+
+    // make a list of DynamicOutputSizes and initialize kwTensor
+    SmallVector<Value> dynOutputSizes;
+    Value kwTensor;
+    for (unsigned i = 0; i < rank - 1; i++) {
+      if (OutputType.isDynamicDim(i)) {
+        dynOutputSizes.push_back(getDimOp(rewriter, loc, input, i));
+      }
+    }
+    if (OutputType.isDynamicDim(rank - 1)) {
+      dynOutputSizes.push_back(HoutIndex);
+      kwTensor = rewriter.create<tensor::EmptyOp>(
+          loc, RankedTensorType::get({ShapedType::kDynamic}, elementType),
+          HoutIndex);
+    } else {
+      kwTensor = rewriter.create<tensor::EmptyOp>(
+          loc,
+          RankedTensorType::get({OutputType.getShape()[rank - 1]}, elementType),
+          ValueRange{});
+    }
```

Comment:
You can use the helpers in

https://github.com/llvm/torch-mlir/blob/be3e74b647375d1c72f075bebf77dc31e74477a2/lib/Conversion/Utils/Utils.cpp#L124-L131

to create tensors without having to worry about dynamic/static cases.

Reply:
I am now using createInitTensor, and the OpFoldResult builder for EmptyOp in a few places now. Very helpful, thank you.